foo/lib/config/configloader.py

```diff
@@ -0,0 +1,98 @@
+import os
+from typing import Optional
+
+import boto3
+import yaml
+
+
+class ConfigLoaderException(Exception):
+    pass
+
+
+class Config:
+
+    def __init__(self, config: dict, source: str):
+        self._config = config
+        self._source = source
+
+    @property
+    def source(self):
+        return self._source
+
+    def __repr__(self):
+        return (f'Config(config={self._config!r}'
+                f'       source={self.source!r})')
+
+    def __getattr__(self, name):
+        return self._config[name]
+
+    def __getitem__(self, name):
+        return self._config[name]
+
+    def get(self, name, default=None):
+        return self._config.get(name, default)
+
+
+class ConfigLoader:
+
+    """Loads configuration files.
+
+    Source of the config depends on the currently set environment (either
+    'test' or 'prod', as dictated by the *env* file.
+    """
+
+    S3_BUCKET = 'configs'
+    _instance = None
+    _instance_ready = False
+
+    # Make this class a singleton since usually config is loaded only once.
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def __init__(self, defaults: dict = {}):
```

Comment:
Do not use mutable elements in function arguments. Change it to:
```suggestion
    def __init__(self, defaults: Optional[Dict] = None):
        if defaults is None:
            defaults = {}
```